test/python/circuit/library/test_linear_function.py

```diff
@@ -0,0 +1,224 @@
+
+        for _ in range(10):
+            # create a random invertible binary matrix
+            binary_matrix = random_invertible_binary_matrix(num_qubits, seed=rng)
+
+            # create a linear function with this matrix
+            linear_function = LinearFunction(binary_matrix, validate_input=True)
+            self.assertTrue(np.all(linear_function.linear == binary_matrix))
+
+            # synthesize linear function
+            synthesized_circuit = linear_function.definition
+            self.assertTrue(isinstance(synthesized_circuit, QuantumCircuit))
+
+            # check that the synthesized linear function only contains CX and SWAP gates
+            for inst, _, _ in synthesized_circuit.data:
+                self.assertTrue(isinstance(inst, (CXGate, SwapGate)))
+
+            # construct a linear function out of this linear circuit
+            synthesized_linear_function = LinearFunction(synthesized_circuit, validate_input=True)
+
+            # check equivalence of the two linear matrices
+            self.assertTrue(np.all(synthesized_linear_function.linear == binary_matrix))
+
+    def test_patel_markov_hayes(self):
+        """Checks the explicit example from Patel-Markov-Hayes's paper."""
+
+        # This code is adapted from test_synthesis.py
+        binary_matrix = [
+            [1, 1, 0, 0, 0, 0],
+            [1, 0, 0, 1, 1, 0],
+            [0, 1, 0, 0, 1, 0],
+            [1, 1, 1, 1, 1, 1],
+            [1, 1, 0, 1, 1, 1],
+            [0, 0, 1, 1, 1, 0],
+        ]
+
+        # Construct linear function from matrix: here, we copy a matrix
+        linear_function_from_matrix = LinearFunction(binary_matrix, validate_input=True)
+
+        # Create the circuit displayed above:
+        linear_circuit = QuantumCircuit(6)
+        linear_circuit.cx(4, 3)
+        linear_circuit.cx(5, 2)
+        linear_circuit.cx(1, 0)
+        linear_circuit.cx(3, 1)
+        linear_circuit.cx(4, 2)
+        linear_circuit.cx(4, 3)
+        linear_circuit.cx(5, 4)
+        linear_circuit.cx(2, 3)
+        linear_circuit.cx(3, 2)
+        linear_circuit.cx(3, 5)
+        linear_circuit.cx(2, 4)
+        linear_circuit.cx(1, 2)
+        linear_circuit.cx(0, 1)
+        linear_circuit.cx(0, 4)
+        linear_circuit.cx(0, 3)
+
+        # Construct linear function from matrix: we build a matrix
+        linear_function_from_circuit = LinearFunction(linear_circuit, validate_input=True)
+
+        # Compare the matrices
+        self.assertTrue(
+            np.all(linear_function_from_circuit.linear == linear_function_from_matrix.linear)
+        )
+
+        self.assertTrue(
+            Operator(linear_function_from_matrix.definition) == Operator(linear_circuit)
+        )
+
+    def test_bad_matrix_non_rectangular(self):
+        """Tests that an error is raised if the matrix is not rectangular."""
+        mat = [[1, 1, 0, 0], [1, 0, 0], [0, 1, 0, 0], [1, 1, 1, 1]]
+        self.assertRaises(CircuitError, LinearFunction, mat)
```

Comment:
Exception assertions can be made a bit more legible with the context-manager form:
```python
with self.assertRaises(CircuitError):
    LinearFunction(mat)
```
or
```python
with self.assertRaisesRegex(CircuitError, "<part of the message you expect"):
    LinearFunction(mat)
```

Reply:
Oh, nice to know! Thanks!